resolver/bootstrap.go

```diff
@@ -112,14 +118,16 @@ func (b *Bootstrap) resolveUpstream(r Resolver, host string) ([]net.IP, error) {
 		cfg := config.GetConfig()
 		ctx := context.Background()
 
-		timeout := cfg.Upstreams.Timeout
-		if timeout.IsAboveZero() {
-			var cancel context.CancelFunc
-
-			ctx, cancel = context.WithTimeout(ctx, timeout.ToDuration())
-			defer cancel()
+		timeout := defaultTimeout
+		if cfg.Upstreams.Timeout.IsAboveZero() {
+			timeout = cfg.Upstreams.Timeout.ToDuration()
 		}
 
+		var cancel context.CancelFunc
+
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
+
```

Comment:
I think we should move the timeout determination to `NewBootstrap` and create `b.dialer` with it.  
That way we can access the potentially user configured timeout via `b.dialer.Timeout` and re-use the dialer as is in `NewHTTPTransport`.

```suggestion
		ctx, cancel := context.WithTimeout(context.Background(), b.dialer.Timeout)
		defer cancel()

```

(Also I think the `var cancel context.CancelFunc` shouldn't be needed even in the previous code and `:=` would work)

Reply:
Yes it reduces some code duplication and we can reuse dialer. Your suggestion with "b.dialer.Timeout" doesn't work, because dialer is an interface and not net.Dialer. I introduced a new field "timeout" to avoid some "hacks" with casting etc.

Reply:
Ah right, thanks for adjusting!